<a href="https://colab.research.google.com/github/Shatha2040/-ALAN-/blob/main/Saudi_Stock_AD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import matplotlib.pyplot as plt



In [ ]:
df = pd.read_csv('/content/Tadawul_stcks_23_4 (2).csv')

# **Data Understanding**

In [ ]:

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6992 entries, 0 to 6991
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   symbol          6992 non-null   int64  
 1   name            6992 non-null   object 
 2   trading_name    6992 non-null   object 
 3   sectoer         6992 non-null   object 
 4   date            6992 non-null   object 
 5   open            6830 non-null   float64
 6   high            6830 non-null   float64
 7   low             6830 non-null   float64
 8   close           6992 non-null   float64
 9   change          6992 non-null   float64
 10  perc_Change     6992 non-null   float64
 11  volume_traded   6992 non-null   float64
 12  value_traded    6992 non-null   float64
 13  no_trades       6992 non-null   float64
dtypes: float64(9), int64(1), object(4)
memory usage: 764.9+ KB


In [ ]:
df

,symbol,name,trading_name,sectoer,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades
0,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-23,31.80,31.80,31.45,31.60,0.25,0.80,230790.0,7295880.55,507.0
1,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-22,30.50,31.40,30.30,31.35,-0.35,-1.10,422695.0,13113020.10,1015.0
2,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-21,31.70,32.30,31.30,31.70,-0.75,-2.31,266845.0,8497563.05,668.0
3,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-20,32.70,32.90,32.10,32.45,-0.25,-0.76,254463.0,8271461.45,557.0
4,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-19,32.50,32.80,32.45,32.70,0.50,1.55,199425.0,6500652.35,498.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6987,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-12,20.20,21.20,20.16,20.70,-1.70,-7.59,1618746.0,33206311.44,2329.0
6988,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-11,24.50,24.50,21.60,22.40,-1.60,-6.67,1487872.0,33903289.20,2092.0
6989,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-10,22.98,24.24,22.68,24.00,1.22,5.36,1877511.0,43740461.26,2359.0
6990,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-09,22.78,23.18,22.78,22.78,-2.52,-9.96,1041240.0,23747279.76,893.0


In [ ]:
df.columns

# تنظيف البيانات:
# - حذف المسافات الزائدة من أسماء الأعمدة
# - تعديل أسماء الأعمدة التالية:
#   trading_name
#   volume_traded
#   no_trades

Index(['symbol', 'name', 'trading_name ', 'sectoer', 'date', 'open', 'high',
       'low', 'close', 'change', 'perc_Change', 'volume_traded ',
       'value_traded', 'no_trades '],
      dtype='object')

In [ ]:
result = pd.DataFrame({
    "Column": df.columns,
    "Sample Value": df.iloc[0].values,
    "Value Type": [type(x).__name__ for x in df.iloc[0]],
    "Data Type": df.dtypes.values
})

print(result)

# تنظيف وتعديل البيانات:

# - تحويل نوع عمود التاريخ إلى مناسب (إذا كان object يحتاج تحويل إلى datetime)
# - تعديل اسم عمود sector ليكون صحيح/موحد
# - حذف عمود name (لعدم الحاجة والاكتفاء بـ trading_name)
# - volume_traded: تحويله إلى int (يمثل عدد الأسهم المتداولة)
# - no_trades: تحويله إلى int (يمثل عدد الصفقات/المعاملات)

            Column                 Sample Value Value Type Data Type
0           symbol                         2030      int64     int64
1             name  Saudi Arabia Refineries Co.        str    object
2    trading_name                         SARCO        str    object
3          sectoer                       Energy        str    object
4             date                   2020-04-23        str    object
5             open                         31.8    float64   float64
6             high                         31.8    float64   float64
7              low                        31.45    float64   float64
8            close                         31.6    float64   float64
9           change                         0.25    float64   float64
10     perc_Change                          0.8    float64   float64
11  volume_traded                      230790.0    float64   float64
12    value_traded                   7295880.55    float64   float64
13      no_trades                 

In [ ]:
#object

df.describe(include='object')

#top	أكثر قيمة تكررت
#freq	عدد مرات تكرار الـ top

,name,trading_name,sectoer,date
count,6992,6992,6992,6992
unique,200,200,11,35
top,Saudi Arabia Refineries Co.,SARCO,Financials,2020-04-23
freq,35,35,1645,200


In [ ]:
df.describe()

#---------------------------------------

# angeIndex: 6992 entries, 0 to 6991
#open : 6830.000000 / high : 6830.000000 / 6992 - 6830 = 162 (NAN)
#Scientific Notation / e+06 = × 10^6




,symbol,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades
count,6992.000000,6830.000000,6830.000000,6830.000000,6992.000000,6992.000000,6992.000000,6.992000e+03,6.992000e+03,6992.000000
mean,4113.914903,24.659725,25.182366,24.196105,24.328474,-0.089798,-0.336230,1.136108e+06,2.358603e+07,1076.635584
std,2257.109245,22.733866,23.249451,22.309276,22.576215,1.266593,3.683778,3.011323e+06,6.681986e+07,1648.871027
min,1010.000000,5.740000,5.950000,5.740000,5.390000,-14.000000,-12.880000,0.000000e+00,0.000000e+00,0.000000
25%,2200.000000,10.360000,10.580000,10.140000,10.220000,-0.340000,-1.850000,1.980355e+05,3.322344e+06,336.000000
50%,4040.000000,15.810000,16.120000,15.440000,15.500000,0.000000,0.000000,4.225855e+05,8.190636e+06,639.500000
75%,5110.000000,29.700000,30.300000,29.050000,29.060000,0.250000,1.480000,9.724018e+05,1.999385e+07,1193.000000
max,8312.000000,157.000000,163.000000,155.400000,158.800000,12.400000,19.930000,6.831751e+07,1.064690e+09,26201.000000


In [ ]:
df[df.duplicated()]

,symbol,name,trading_name,sectoer,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades


In [ ]:
df.isnull().sum()
#open
#high
#low

,0
symbol,0
name,0
trading_name,0
sectoer,0
date,0
open,162
high,162
low,162
close,0
change,0


# **Null Analysis**

In [ ]:
df[df.isnull().any(axis=1)]

"""
السؤال: التعامل مع القيم المفقودة (NaN)

- هل يتم حذف القيم المفقودة أم استبدالها؟ أم يتم إنشاء Flag؟

- عدد القيم المفقودة: 162 من إجمالي 6992

- هل القيم المفقودة موزعة على الصفوف بشكل جزئي؟
  أم أن الثلاثة أعمدة تكون NaN في نفس الصف؟

- هل القيم المفقودة مرتبطة بشركة معينة أو قطاع محدد؟

الحالات:

الحالة الأولى:
open, high, low, close, change, perc_Change موجودة
volume_traded, value_traded, no_trades = 0 أو NaN
→ يعني لا يوجد نشاط تداول أو بيانات غير مكتملة

الحالة الثانية:
volume_traded, value_traded, no_trades > 0
open, high, low = NaN
→ يوجد تداول فعلي لكن بيانات الأسعار مفقودة
"""



'\nالسؤال: التعامل مع القيم المفقودة (NaN)\n\n- هل يتم حذف القيم المفقودة أم استبدالها؟ أم يتم إنشاء Flag؟\n\n- عدد القيم المفقودة: 162 من إجمالي 6992\n\n- هل القيم المفقودة موزعة على الصفوف بشكل جزئي؟\n  أم أن الثلاثة أعمدة تكون NaN في نفس الصف؟\n\n- هل القيم المفقودة مرتبطة بشركة معينة أو قطاع محدد؟\n\nالحالات:\n\nالحالة الأولى:\nopen, high, low, close, change, perc_Change موجودة\nvolume_traded, value_traded, no_trades = 0 أو NaN\n→ يعني لا يوجد نشاط تداول أو بيانات غير مكتملة\n\nالحالة الثانية:\nvolume_traded, value_traded, no_trades > 0\nopen, high, low = NaN\n→ يوجد تداول فعلي لكن بيانات الأسعار مفقودة\n'

# **التعامل مع الحالة الآولى**

In [ ]:
print(f"اجمالي عدد القطاعات  {df['sectoer'].nunique()}")
print("*"*100)

null_counts_S = df[df[['open','high','low']].isnull().all(axis=1)].groupby('sectoer').size()

total_counts_S = df.groupby('sectoer').size()

result = pd.DataFrame({
    'total': total_counts_S,
    'null_open_high_low': null_counts_S
}).fillna(0)

result['null_ratio'] = result['null_open_high_low'] / result['total']

print(result.sort_values('null_open_high_low', ascending=False))


اجمالي عدد القطاعات  11
****************************************************************************************************
                        total  null_open_high_low  null_ratio
sectoer                                                      
Consumer Staples          560                39.0    0.069643
Financials               1645                36.0    0.021884
Communication Services    210                35.0    0.166667
Industrials               700                35.0    0.050000
Real Estate               980                14.0    0.014286
Information Technology     69                 2.0    0.028986
Materials                1470                 1.0    0.000680
Consumer Discretionary    840                 0.0    0.000000
Energy                    175                 0.0    0.000000
Health Care               273                 0.0    0.000000
Utilities                  70                 0.0    0.000000



الحالة الأولى: بعد تحليل القيم المفقودة في الشركات، تم التوصل إلى الاستنتاج التالي:

المشكلة ليست في القطاع نفسه،
بل في شركات محددة داخل كل قطاع.

Communication Services:
- جميع القيم المفقودة تقريبًا (35) تعود إلى شركة ATHEEB TELECOM

➡️ هذا لا يعني وجود مشكلة في القطاع،
بل يشير إلى أن الشركة كانت متوقفة عن التداول خلال الفترة.

Industrials:
- جميع القيم المفقودة (35) تعود إلى شركة ALKHODARI

Financials:
- القيم المفقودة غالبًا تعود إلى شركة WAFA INSURANCE

📌 الاستنتاج:
القيم المفقودة مرتبطة بشركات محددة متوقفة عن التداول،
وليست مشكلة على مستوى القطاعات.




----------------------------------------------------------------------------
No Trading Activity

كل الأربع شركات عندها نفس النمط:

open    high    low     close   volume  no_trades
NaN     NaN     NaN     ثابت     0       0

يعني:
لم يحدث أي تداول خلال اليوم، لذلك:
- لا يوجد سعر افتتاح (open)
- لا يوجد أعلى سعر (high)
- لا يوجد أقل سعر (low)

لكن:
- سعر الإغلاق (close) يبقى ثابت لأنه يمثل آخر سعر معروف قبل توقف التداول

📌 الخلاصة:
القيم المفقودة هنا ليست فقدان بيانات عشوائي،
بل تعكس حالة "No Trading Activity" بشكل منطقي وواضح.

✔️ لذلك:
NULL له معنى في البيانات
"""

In [ ]:
print(f" اجمالي عدد الشركات: {df['trading_name '].nunique()}")
print("*"*100)

null_counts_T = df[df[['open','high','low']].isnull().all(axis=1)].groupby('trading_name ').size()

total_counts_T = df.groupby('trading_name ').size()

result = pd.DataFrame({
    'total': total_counts_T,
    'null_open_high_low': null_counts_T
}).fillna(0)

result['null_ratio'] = result['null_open_high_low'] / result['total']

#print(result.sort_values('null_open_high_low', ascending=False))

print(result.sort_values('null_open_high_low', ascending=False).to_string())


 اجمالي عدد الشركات: 200
****************************************************************************************************
                       total  null_open_high_low  null_ratio
trading_name                                                
ATHEEB TELECOM            35                35.0    1.000000
ALKHODARI                 35                35.0    1.000000
THIMAR                    35                35.0    1.000000
WAFA INSURANCE            35                35.0    1.000000
MULKIA REIT               35                 4.0    0.114286
BONYAN REIT               35                 2.0    0.057143
ALKHABEER REIT            35                 2.0    0.057143
ARAB SEA                  34                 2.0    0.058824
SWICORP WABEL REIT        35                 2.0    0.057143
TADCO                     35                 2.0    0.057143
WAFRAH                    35                 2.0    0.057143
MASHAAR REIT              35                 1.0    0.028571
ENAYA               

📊 تحليل القيم المفقودة حسب الشركات

| Trading Name   | Total | Null Open/High/Low | Null Ratio |
| -------------- | ----: | -----------------: | ---------: |
| ATHEEB TELECOM |    35 |               35.0 |   1.000000 |
| ALKHODARI      |    35 |               35.0 |   1.000000 |
| THIMAR         |    35 |               35.0 |   1.000000 |
| WAFA INSURANCE |    35 |               35.0 |   1.000000 |

📌 الملاحظة:
القيم Null (35) تمثل عدد الأيام التي لا يوجد فيها نشاط تداول لهذه الشركات خلال الفترة.

❌ لذلك لا يمكن حذف هذه الصفوف، لأن الحذف سيعطي انطباعًا خاطئًا بأن الشركات غير موجودة في السوق خلال تلك الأيام.

❓ القرار:
هل يتم الاستبدال أم إنشاء Flag؟

⚠️ لماذا لا يتم الاستبدال (Mean Imputation)؟
- سيؤدي إلى توليد أسعار غير حقيقية
- يشوه البيانات المالية الفعلية
- غير منطقي في بيانات السوق لأن الأسعار لا تُخترع
- لو بنستبدل ٠ ، NaN ≠ 0 له معنى اخر

✔️ القرار النهائي:
تم اختيار إنشاء Flag بدلًا من الحذف أو الاستبدال.

📌 مبرر القرار:
بعد تحليل القيم المفقودة في الأعمدة السعرية (open, high, low)، تبيّن أن هذه القيم لم تكن ناتجة عن فقدان عشوائي، بل عن حالات "No Trading Activity"، حيث:
- volume_traded = 0
- no_trades = 0
- close = ثابت (آخر سعر معروف)

🔍 التصنيف:
Missing by Business Logic

أي أن القيم المفقودة ناتجة عن طبيعة السوق وليس خطأ في البيانات.

✔️ النتيجة:

تم الاحتفاظ بالبيانات كما هي، مع إضافة متغير (Flag) لتحديد أيام عدم وجود التداول،
لاستخدامه لاحقًا في التحليل دون تشويه البيانات الأصلية.

اضافه عامود No Trading Activity. = عشان برضو في الشارت استبعده  اذا احتجت




---




Missing Value Handling Decision




| Aspect                               | Mean Imputation                                                                                                                                                                                               | Flag Approach (Selected Method)                                                                                                                                        |
| ------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Data Integrity                       | May create unrealistic price values                                                                                                                                                                           | Preserves original data                                                                                                                                                |
| Business Meaning                     | Ignores the trading context                                                                                                                                                                                   | Preserves the meaning of no trading activity                                                                                                                           |
| Financial Analysis Accuracy          | May distort financial results                                                                                                                                                                                 | Maintains analysis accuracy                                                                                                                                            |
| Impact on Models                     | May introduce bias into models                                                                                                                                                                                | Reduces misleading patterns                                                                                                                                            |
| Interpretability                     | Lower interpretability                                                                                                                                                                                        | Higher interpretability                                                                                                                                                |
| Performance                          | Requires additional computations to calculate replacement values and inject artificial data into the dataset, which may increase processing complexity and affect statistical calculations and model behavior | More efficient because it preserves the original data structure and only adds a lightweight Boolean flag (`True/False`) with minimal memory and computational overhead |
| Suitability for Financial Price Data | ❌ Not suitable                                                                                                                                                                                                | ✅ Suitable                                                                                                                                                             |
| Handling Type                        | Statistical Imputation                                                                                                                                                                                        | Context-Aware Handling                                                                                                                                                 |
| Final Decision                       | Not Selected                                                                                                                                                                                                  | ✅ Selected                                                                                                                                                             |


In [ ]:
df['no_trading_activity'] = (
    df[['open', 'high', 'low']].isnull().all(axis=1)
) & (
    (df['volume_traded '] == 0) &
    (df['no_trades '] == 0)
)

df

,symbol,name,trading_name,sectoer,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades,no_trading_activity
0,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-23,31.80,31.80,31.45,31.60,0.25,0.80,230790.0,7295880.55,507.0,False
1,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-22,30.50,31.40,30.30,31.35,-0.35,-1.10,422695.0,13113020.10,1015.0,False
2,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-21,31.70,32.30,31.30,31.70,-0.75,-2.31,266845.0,8497563.05,668.0,False
3,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-20,32.70,32.90,32.10,32.45,-0.25,-0.76,254463.0,8271461.45,557.0,False
4,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-19,32.50,32.80,32.45,32.70,0.50,1.55,199425.0,6500652.35,498.0,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6987,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-12,20.20,21.20,20.16,20.70,-1.70,-7.59,1618746.0,33206311.44,2329.0,False
6988,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-11,24.50,24.50,21.60,22.40,-1.60,-6.67,1487872.0,33903289.20,2092.0,False
6989,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-10,22.98,24.24,22.68,24.00,1.22,5.36,1877511.0,43740461.26,2359.0,False
6990,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-09,22.78,23.18,22.78,22.78,-2.52,-9.96,1041240.0,23747279.76,893.0,False


📌 الحالة الثانية: (Missing Prices with Active Trading)


يوجد تداول فعلي لكن البيانات ناقصة=التداول حصل فعلًا

open/high/low = NaN


لكن:
volume_traded > 0
أو no_trades > 0

طيب بحلل النشاط وسعر
هل لو حذفت ١٥ علما بان بتمركز قي قطاع واحد هذا ١٥ هل بياثر في النشاط ؟

15 / 6992 ≈ 0.21% يعني التاثير يعني عادي

1) على تحليل السعر (Price Analysis)

مايفرق لان مو مستخدمتها كلهم نل


2) على تحليل النشاط (Volume / Trades)

0.21% هنا لو حذفت بخسر بنسبه

بس المشكله انه متمركز على قطاع معين الي هو

Real Estate فقط (متركّز في قطاع واحد)

وموزعه على الشركات مو شركه معينه في القطاع
--------------------------
15 / 980 ≈ 1.53%

حذف 15 صف من 980 في القطاع = تأثير بسيط جدًا (~1.5%) على النشاط، وغالبًا ما يغيّر النتائج العامة

In [ ]:
to_delete = df[
    df["open"].isna() &
    df["high"].isna() &
    df["low"].isna() &
    (df["volume_traded "] > 0)
]

print(to_delete.shape)
to_delete

(15, 15)


,symbol,name,trading_name,sectoer,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades,no_trading_activity
634,2150,The National Company for Glass Industries,ZOUJAJ,Materials,2020-04-19,NaN,NaN,NaN,14.02,0.0,0.0,14260.0,200266.32,81.0,False
6142,4333,Taleem REIT Fund,TALEEM REIT,Real Estate,2020-03-19,NaN,NaN,NaN,10.76,0.0,0.0,8199.0,88216.90,61.0,False
6223,4336,Mulkia Gulf Real Estate REIT,MULKIA REIT,Real Estate,2020-04-22,NaN,NaN,NaN,7.99,0.0,0.0,11579.0,92545.28,96.0,False
6225,4336,Mulkia Gulf Real Estate REIT,MULKIA REIT,Real Estate,2020-04-20,NaN,NaN,NaN,8.15,0.0,0.0,6197.0,50255.34,70.0,False
6226,4336,Mulkia Gulf Real Estate REIT,MULKIA REIT,Real Estate,2020-04-19,NaN,NaN,NaN,8.15,0.0,0.0,13176.0,107121.47,111.0,False
6242,4336,Mulkia Gulf Real Estate REIT,MULKIA REIT,Real Estate,2020-03-26,NaN,NaN,NaN,8.10,0.0,0.0,28929.0,236375.35,211.0,False
6275,4337,Al Mashaar REIT,MASHAAR REIT,Real Estate,2020-03-30,NaN,NaN,NaN,6.70,0.0,0.0,11818.0,79701.75,89.0,False
6292,4338,AlAhli REIT Fund 1,ALAHLI REIT 1,Real Estate,2020-04-23,NaN,NaN,NaN,7.84,0.0,0.0,11756.0,92161.19,114.0,False
6433,4344,SEDCO Capital REIT Fund,SEDCO CAPITAL REIT,Real Estate,2020-04-22,NaN,NaN,NaN,8.01,0.0,0.0,6070.0,48944.39,66.0,False
6475,4345,Swicorp Wabel REIT Fund,SWICORP WABEL REIT,Real Estate,2020-04-13,NaN,NaN,NaN,7.00,0.0,0.0,10654.0,74541.09,61.0,False


In [ ]:
df = df[
    ~(
        df["open"].isna() &
        df["high"].isna() &
        df["low"].isna() &
        (df["volume_traded "] > 0)
    )
]



In [ ]:
to_delete = df[
    df["open"].isna() &
    df["high"].isna() &
    df["low"].isna() &
    (df["volume_traded "] > 0)
]

print(to_delete.shape)

(0, 15)


# **Data Standardization** -----------------------

**Renaming Columns**

In [ ]:
df = df.rename(columns={
    "trading_name ": "company_name",
    "volume_traded ": "volume_traded",
    "no_trades ": "no_trades",
    "sectoer": "sector",
})
df

,symbol,name,company_name,sector,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades,no_trading_activity
0,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-23,31.80,31.80,31.45,31.60,0.25,0.80,230790.0,7295880.55,507.0,False
1,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-22,30.50,31.40,30.30,31.35,-0.35,-1.10,422695.0,13113020.10,1015.0,False
2,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-21,31.70,32.30,31.30,31.70,-0.75,-2.31,266845.0,8497563.05,668.0,False
3,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-20,32.70,32.90,32.10,32.45,-0.25,-0.76,254463.0,8271461.45,557.0,False
4,2030,Saudi Arabia Refineries Co.,SARCO,Energy,2020-04-19,32.50,32.80,32.45,32.70,0.50,1.55,199425.0,6500652.35,498.0,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6987,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-12,20.20,21.20,20.16,20.70,-1.70,-7.59,1618746.0,33206311.44,2329.0,False
6988,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-11,24.50,24.50,21.60,22.40,-1.60,-6.67,1487872.0,33903289.20,2092.0,False
6989,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-10,22.98,24.24,22.68,24.00,1.22,5.36,1877511.0,43740461.26,2359.0,False
6990,4321,Arabian Centres Co.,ALMRAKEZ,Real Estate,2020-03-09,22.78,23.18,22.78,22.78,-2.52,-9.96,1041240.0,23747279.76,893.0,False


In [ ]:
df['volume_traded']

,volume_traded
0,230790.0
1,422695.0
2,266845.0
3,254463.0
4,199425.0
...,...
6987,1618746.0
6988,1487872.0
6989,1877511.0
6990,1041240.0


In [ ]:
df['no_trades']


,no_trades
0,507.0
1,1015.0
2,668.0
3,557.0
4,498.0
...,...
6987,2329.0
6988,2092.0
6989,2359.0
6990,893.0


**Data Type Conversion**

In [ ]:
df["date"] = pd.to_datetime(df["date"])
df["volume_traded"] = df["volume_traded"].astype(int)
df["no_trades"] = df["no_trades"].astype(int)

In [ ]:
df.dtypes

,0
symbol,int64
name,object
company_name,object
sector,object
date,datetime64[ns]
open,float64
high,float64
low,float64
close,float64
change,float64


**Column Removal**

In [ ]:
df.drop("name", axis=1, inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6977 entries, 0 to 6991
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   symbol               6977 non-null   int64         
 1   company_name         6977 non-null   object        
 2   sector               6977 non-null   object        
 3   date                 6977 non-null   datetime64[ns]
 4   open                 6830 non-null   float64       
 5   high                 6830 non-null   float64       
 6   low                  6830 non-null   float64       
 7   close                6977 non-null   float64       
 8   change               6977 non-null   float64       
 9   perc_Change          6977 non-null   float64       
 10  volume_traded        6977 non-null   int64         
 11  value_traded         6977 non-null   float64       
 12  no_trades            6977 non-null   int64         
 13  no_trading_activity  6977 non-null   b

#**Exploratory Data Analysis (EDA)**

Data Visualization

**4.1 Trading Activity Analysis**

In [ ]:
df.head(5)

,symbol,company_name,sector,date,open,high,low,close,change,perc_Change,volume_traded,value_traded,no_trades,no_trading_activity
0,2030,SARCO,Energy,2020-04-23,31.8,31.8,31.45,31.60,0.25,0.80,230790,7295880.55,507,False
1,2030,SARCO,Energy,2020-04-22,30.5,31.4,30.30,31.35,-0.35,-1.10,422695,13113020.10,1015,False
2,2030,SARCO,Energy,2020-04-21,31.7,32.3,31.30,31.70,-0.75,-2.31,266845,8497563.05,668,False
3,2030,SARCO,Energy,2020-04-20,32.7,32.9,32.10,32.45,-0.25,-0.76,254463,8271461.45,557,False
4,2030,SARCO,Energy,2020-04-19,32.5,32.8,32.45,32.70,0.50,1.55,199425,6500652.35,498,False


1-Top Trading Volume Companies

In [ ]:
top_volume = (
    df.groupby("company_name")["volume_traded"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

fig = px.bar(
    x=top_volume.index,
    y=top_volume.values,
    title='Top 10 Companies by Trading Volume',
    labels={'x': 'Company', 'y': 'Trading Volume'}
)

fig.update_layout(template="plotly_dark")


2-Top Trading Value Companies

In [ ]:
top_value = (
    df.groupby("company_name")["value_traded"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

fig = px.bar(
    x=top_value.index,
    y=top_value.values,
    title='Top 10 Companies by Trading Value',
    labels={'x': 'Company', 'y': 'Trading Value'}
)

#fig.show()
fig.update_layout(template="plotly_dark")


3-Trading Value by Sector

In [ ]:
sector_value = (
    df.groupby("sector")["value_traded"]
    .sum()
    .sort_values(ascending=False)
)

fig = px.pie(
    names=sector_value.index,
    values=sector_value.values,
    title='Trading Value by Sector',
    hole=0.2
)
fig.update_layout(template="plotly_dark")



4- Number of Trades Analysis


In [ ]:

fig = go.Figure()

# Companies data
companies = (
    df.groupby("company_name")["no_trades"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

# Sectors data
sectors = (
    df.groupby("sector")["no_trades"]
    .sum()
    .sort_values(ascending=False)
)

# الشكل الأول (Companies)
fig.add_trace(go.Bar(
    x=companies.index,
    y=companies.values,
    name="Companies"
))

# الشكل الثاني (Sectors)
fig.add_trace(go.Bar(
    x=sectors.index,
    y=sectors.values,
    name="Sectors",
    visible=False
))

# Dropdown Menu
fig.update_layout(
    updatemenus=[
        dict(
            type="dropdown",
            buttons=[
                dict(
                    label="Companies",
                    method="update",
                    args=[{"visible": [True, False]},
                          {"title": "Number of Trades by Companies"}]
                ),
                dict(
                    label="Sectors",
                    method="update",
                    args=[{"visible": [False, True]},
                          {"title": "Number of Trades by Sectors"}]
                )
            ]
        )
    ],
    title="Number of Trades Analysis"
)
fig.update_layout(template="plotly_dark")


5-Non-Trading Activity Analysis

In [ ]:
no_trading_count = df[df["no_trading_activity"] == True].shape[0]
no_trading_count

147

In [ ]:
no_trading_ratio = df["no_trading_activity"].mean() * 100
#no_trading_ratio
# np.float64(2.106922746165974) = 2.1%
print(f"No Trading Activity Ratio: {no_trading_ratio:.2f}%")

No Trading Activity Ratio: 2.11%


In [ ]:

stagnant_companies = (
    df[df["no_trading_activity"] == True]
    .groupby("company_name")
    .size()
    .sort_values(ascending=False)
    .head(10)
)

fig = px.bar(
    x=stagnant_companies.index,
    y=stagnant_companies.values,
    title="Most Inactive Companies (No Trading Activity)",
    labels={"x": "Company", "y": "Inactive Days"}
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


In [ ]:
stagnant_sectors = (
    df[df["no_trading_activity"] == True]
    .groupby("sector")
    .size()
    .sort_values(ascending=False)
)

fig = px.bar(
    x=stagnant_sectors.values,
    y=stagnant_sectors.index,
    orientation='h',
    title="Inactive Sectors (No Trading Activity)",
    labels={"x": "Inactive Days", "y": "Sector"}
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


6-Activity Spike Analysis (Volume + Value)

حساب Z-Score للـ Volume و Value

هنا باخذها على مستويين ⁉

Company-level:

هل الشركة خرجت عن نمطها المعتاد؟

Sector-level:

هل الشركة تجاوزت سلوك قطاعها؟



 لو استخدمنا merge بنحتاج نسوي جدول جديد

 نحسب فيه mean و std لكل شركة،

وبعدها نرجع ندمجه مع الداتا الأصلية.

لكن باستخدام transform يطبق الحسابات مباشرة

 على كل شركة ويرجع النتائج بنفس عدد الصفوف،

 فصار أبسط وأنظف لحساب الـ Z-score.

In [ ]:
df["volume_z"] = df.groupby("company_name")["volume_traded"].transform( lambda x: (x - x.mean()) / x.std() )

In [ ]:
df["volume_spike"] = df["volume_z"] > 2

In [ ]:
df["volume_spike"].value_counts()

,count
volume_spike,
False,6607
True,370


In [ ]:
volume_spike_companies = (
    df[df["volume_spike"] == True]
    .groupby("company_name")
    .size()
    .sort_values(ascending=False)
    .head(15)
)

volume_spike_companies

,0
company_name,
BATIC,4
ALHAMMADI,4
AICC,4
SADAFCO,4
SAMBA,4
MAHARAH,4
ALKHALEEJ TRNG,3
ALASEEL,3
Al RAJHI REIT,3


In [ ]:
import plotly.express as px

fig = px.box(
    df,
    y="volume_z",
    title="Volume Z-Score Distribution"
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


In [ ]:

fig = px.bar(
    x=volume_spike_companies.index,
    y=volume_spike_companies.values,
    title="Top Companies with Volume Spikes",
    labels={
        "x": "Company",
        "y": "Spike Count"
    }
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


In [ ]:
df["value_z"] = df.groupby("company_name")["value_traded"].transform(lambda x: (x - x.mean()) / x.std())

df["value_spike"] = df["value_z"] > 2

In [ ]:
df["value_spike"].value_counts()

,count
value_spike,
False,6613
True,364


In [ ]:
value_spike_companies = (
    df[df["value_spike"] == True]
    .groupby("company_name")
    .size()
    .sort_values(ascending=False)
    .head(15)
)

value_spike_companies

,0
company_name,
BATIC,4
SAMBA,4
AICC,3
ALKHALEEJ TRNG,3
ALRAJHI TAKAFUL,3
ARAB SEA,3
AL MAATHER REIT,3
ALASEEL,3
ALKATHIRI,3


In [ ]:

fig = px.bar(
    x=value_spike_companies.index,
    y=value_spike_companies.values,
    title="Top Companies with Value Spikes",
    labels={
        "x": "Company",
        "y": "Spike Count"
    }
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


In [ ]:
both_spikes = (
    df[
        (df["volume_spike"] == True) &
        (df["value_spike"] == True)
    ]
    .groupby("company_name")
    .size()
    .sort_values(ascending=False)
)

both_spikes

,0
company_name,
BATIC,4
SAMBA,4
Al RAJHI REIT,3
ALHAMMADI,3
AICC,3
...,...
WALAA,1
YCC,1
WATANIYA,1


In [ ]:
fig = px.bar(
    x=both_spikes.index,
    y=both_spikes.values,
    title="Companies with Both Volume and Value Spikes",
    labels={
        "x": "Company",
        "y": "Both Spike Count"
    }
)

fig.update_layout(title_x=0.5)
fig.update_layout(template="plotly_dark")


7-Trading Activity Over Time

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6977 entries, 0 to 6991
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   symbol               6977 non-null   int64         
 1   company_name         6977 non-null   object        
 2   sector               6977 non-null   object        
 3   date                 6977 non-null   datetime64[ns]
 4   open                 6830 non-null   float64       
 5   high                 6830 non-null   float64       
 6   low                  6830 non-null   float64       
 7   close                6977 non-null   float64       
 8   change               6977 non-null   float64       
 9   perc_Change          6977 non-null   float64       
 10  volume_traded        6977 non-null   int64         
 11  value_traded         6977 non-null   float64       
 12  no_trades            6977 non-null   int64         
 13  no_trading_activity  6977 non-null   b

In [ ]:
df = df.set_index("date")

In [ ]:
daily = df.resample("D").agg({
    "volume_traded": "sum",
    "value_traded": "sum",
    "no_trades": "sum"
})

In [ ]:
fig = go.Figure()


fig.add_trace(go.Scatter(
    x=daily.index,
    y=daily["volume_traded"],
    mode="lines",
    name="Volume Traded"
))

fig.add_trace(go.Scatter(
    x=daily.index,
    y=daily["value_traded"],
    mode="lines",
    name="Value Traded",
    visible=False
))

fig.add_trace(go.Scatter(
    x=daily.index,
    y=daily["no_trades"],
    mode="lines",
    name="Number of Trades",
    visible=False
))

fig.update_layout(
    title="Trading Activity Over Time",
    xaxis_title="Date",
    template="plotly_white",

    updatemenus=[
        dict(
            type="dropdown",
            direction="down",
            x=1.1,
            y=1.15,
            showactive=True,

            buttons=[

                dict(
                    label="Volume",
                    method="update",
                    args=[
                        {"visible": [True, False, False]},
                        {"title": "Volume Traded Over Time"}
                    ]
                ),

                dict(
                    label="Value",
                    method="update",
                    args=[
                        {"visible": [False, True, False]},
                        {"title": "Value Traded Over Time"}
                    ]
                ),

                dict(
                    label="No. Trades",
                    method="update",
                    args=[
                        {"visible": [False, False, True]},
                        {"title": "Number of Trades Over Time"}
                    ]
                )
            ]
        )
    ]
)
fig.update_layout(template="plotly_dark")

4.2 Price Analysis

تحليل الأسعار والتغيرات السوقية

تذكيييير استبعاد ⏰
no_trading_activity = True



In [ ]:
price_df = df[~df['no_trading_activity']]

1-Price Trend Over Time

In [ ]:
price_trend = (
    price_df.groupby('date')['close']
    .mean()
    .reset_index()
)

In [ ]:
fig = px.line(
    price_trend,
    x='date',
    y='close',
    title='Average Closing Price Over Time',
    labels={
        'date': 'Date',
        'close': 'Average Closing Price'
    }
)

fig.update_layout(template="plotly_dark")

2-Open vs Close Price Analysis

In [ ]:
open_close_trend = (
    price_df.groupby('date')[['open', 'close']]
    .mean()
    .reset_index()
)

In [ ]:
fig = px.line(
    open_close_trend,
    x='date',
    y=['open', 'close'],
    title='Open vs Close Price Trend Over Time',
    labels={
        'value': 'Average Price',
        'date': 'Date',
        'variable': 'Price Type'
    }
)

fig.update_layout(template="plotly_dark")

In [ ]:
fig = px.scatter(
    price_df,
    x='open',
    y='close',
    color='sector',
    title='Open vs Close Price Analysis',
    labels={
        'open': 'Opening Price',
        'close': 'Closing Price'
    }
)

fig.update_layout(template="plotly_dark")

3-Average Stock Price by Sector

In [ ]:
sector_price = (
    price_df.groupby('sector')['close']
    .mean()
    .sort_values(ascending=False)
    .reset_index()
)

In [ ]:
fig = px.bar(
    sector_price,
    x='sector',
    y='close',
    title='Average Stock Price by Sector',
    labels={
        'sector': 'Sector',
        'close': 'Average Closing Price'
    }
)

fig.update_layout(template="plotly_dark")

4-Highest Percentage Change Companies

Top Gainers

In [ ]:
top_change = (
    price_df.groupby("company_name")['perc_Change']
    .mean()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
)

In [ ]:
fig = px.bar(
    top_change,
    x='company_name',
    y='perc_Change',
    title='Top Companies by Percentage Change',
    labels={
        'trading_name': 'Company',
        'perc_Change': 'Percentage Change'

    }
)

fig.update_layout(template="plotly_dark")

Largest Price Movements

In [ ]:
top_change = (
    price_df.groupby("company_name")['perc_Change']
    .mean()
    .abs()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
)

In [ ]:
fig = px.bar(
    top_change,
    x='perc_Change',
    y='company_name',
    orientation='h',
    title='Top Companies by Percentage Change',
    labels={
        'perc_Change': 'Average Percentage Change',
        'company_name': 'Company'
    }
)


fig.update_layout(template="plotly_dark")

5-Bearish Sector Analysis

In [ ]:
bearish_df = price_df[price_df['close'] < price_df['open']]

bearish_sector = (
    bearish_df.groupby('sector')
    .size()
    .sort_values(ascending=False)
    .reset_index(name='count')
)

In [ ]:
fig = px.bar(
    bearish_sector,
    x='sector',
    y='count',
    title='Bearish Sector Analysis',
    labels={
        'sector': 'Sector',
        'count': 'Number of Bearish Days'
    }
)

fig.update_layout(template="plotly_dark")

6-Volatility Analysis

In [ ]:
volatility = (
    price_df.groupby("company_name")['perc_Change']
    .std()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
)

In [ ]:
fig = px.bar(
    volatility,
    x='company_name',
    y='perc_Change',
    title='Most Volatile Companies',
    labels={
        'company_name': 'Company',
        'perc_Change': 'Volatility'
    }
)

fig.update_layout(template="plotly_dark")

3-Sector Analysis

Sector Trading Activity

إجمالي سيولة القطاع

In [ ]:
sector_activity = (
    df.groupby('sector')['value_traded']
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)

In [ ]:
fig = px.bar(
    sector_activity,
    x='value_traded',
    y='sector',
    orientation='h',
    title='Sector Trading Activity',
    labels={
        'value_traded': 'Total Traded Value',
        'sector': 'Sector'
    }
)

fig.update_layout(template="plotly_dark")

Average Traded Value per Company

هل شركات القطاع نفسها نشطة فعلًا؟

In [ ]:
avg_sector_activity = (
    df.groupby('sector')['value_traded']
    .mean()
    .sort_values(ascending=False)
    .reset_index()
)

In [ ]:
fig = px.bar(
    avg_sector_activity,
    x='value_traded',
    y='sector',
    orientation='h',
    title='Average Traded Value per Company by Sector',
    labels={
        'value_traded': 'Average Traded Value',
        'sector': 'Sector'
    }
)

fig.update_layout(template="plotly_dark")

Financials قطاع ضخم وفيه شركات كثيرة.

لكن Energy كشف أن:

شركاته نفسها قوية تداوليا.

Sector Market Performance

In [ ]:
sector_performance = (
    price_df.groupby('sector')['perc_Change']
    .mean()
    .sort_values(ascending=False)
    .reset_index()
)

In [ ]:
fig = px.bar(
    sector_performance,
    x='perc_Change',
    y='sector',
    orientation='h',
    title='Sector Market Performance',
    labels={
        'perc_Change': 'Average Percentage Change',
        'sector': 'Sector'
    }
)

fig.update_layout(template="plotly_dark")